st2common/st2common/bootstrap/base.py

```diff
@@ -149,13 +152,38 @@ def _register_pack_db(self, pack_name, pack_dir):
         if not content:
             raise ValueError('Pack "%s" metadata file is empty' % (pack_name))
 
-        content['ref'] = pack_name
+        # The rules for the pack ref are as follows:
+        # 1. If ref attribute is available, we used that
+        # 2. If pack_name is available we use that (this only applies to packs
+        # 2hich are in sub-directories)
+        # 2. If attribute is not available, but pack name is and pack name meets the valid name
+        # criteria, we use that
+        if content.get('ref', None):
+            content['ref'] = content['ref']
+        elif re.match(PACK_NAME_WHITELIST_REGEX, pack_name):
+            content['ref'] = pack_name
+        else:
+            if re.match(PACK_NAME_WHITELIST_REGEX, content['name']):
+                content['ref'] = content['name']
```

Comment:
I would prefer to avoid additional magic though - one option would be to update the `ref` regex and only allow lowercase letters.

And this way if pack name is not lowercase already it's up to the user to add an explicit ref :)